designs/workflow-refactor.md

```diff
@@ -0,0 +1,296 @@
+
+We will remove the responsibility of installing all necessary components specific to a provider to providers. The installation will occur during the *Install Bootstrap Components* and *Install Management Components* outlined earlier in this document. Providers will be expected to install and upgrade their infrastructure provider along with any other components.
+
+#### Impact
+
+* A clearer delineation in responsibilities for installing provider specific cluster management components (for example the Tink stack).
+* A clear opportunity for providers to install components they require for provisioning when creating or upgrading clusters.
+* Removal of the `EnvMap` and, possibly, `Version()` methods on the provider interface. These methods, when taken at face value, are examples of ambiguous methods that require reverse engineering to understand how they’re used.
+
+### Workflow hooks
+
+There are a handful of behaviors on the provider interface that suggest they’re run pre or post a workflow task or activity within a task. The unclear expectations on pre/post tasks creates ambiguity and, consequently, hard to rationalize expectations for provider developers and hard for workflow maintainers to know where to instrument pre and post capabilities.
+
+We will introduce hooks that provider implementers can use to execute arbitrary functionality around workflow tasks. Constructs needing to hook into workflows will receive a construct they can use to register handlers. Some hooks will require data unavailable until runtime such as a client configuration for contacting the bootstrap cluster API server. This will be made available via an injected context using well documented helper functions.
+
+#### Example implementation from a hook implementers perspective
+
+```go
+package hook
+
+// Handler is the signature of handler functions invoked before or after a 
+// particular task.
+type Handler func(context.Context) error
+
+// Binder defines methods for binding before or after a given task.
+type Binder interface {
+  // Before registers a Handler to be invoked before Task is executed.
+  Before(Task, Handler) error
+  
+  // After registers a Handler to be invoked after a Task is executed.
+  After(Task, Handler) error
+}
+```
+
+```go
+package foo
+
+import (
+  "github.com/aws/eks-anywhere/pkg/workflow/hook"
+  "github.com/aws/eks-anywhere/pkg/workflow/management"
+  "github.com/aws/eks-anywhere/pkg/workflow/workload"
+)
+
+// Foo is a data structure needing to hook into workflow runs.
+type Foo struct {
+     // Embed a nooping hook registrar to avoid empty method definitions
+    // in providers.
+    workload.NoopHookRegistrar
+}
+
+// BindCreateWorkflowHooks is called from the management.NewCreateWorkflow()
+// func extending an opportunity to hook into a management cluster creation 
+// workflow
+func (Foo) BindCreateWorkflowHooks(binder hook.Binder) {
```

Comment:
How can we keep these bound hooks 
a) easily testable
b) clearly readable?

Reply:
I'm not sure I follow, hooks are just functions in this example. The function could be anonymous (which isn't easily tested) or provided by a high-order function or just an implementation registered here. The latter 2 offer good test-ability in conjunction with documentation on what to expect in the context and any helpers to populate the context.